src/Operation/Unfold.php

```diff
@@ -25,19 +25,19 @@ final class Unfold extends AbstractOperation
     /**
      * @pure
      *
-     * @return Closure(T...): Closure(callable(mixed|T...): (mixed|array<TKey, T>)): Closure(): Generator<int, T>
+     * @return Closure(T...): Closure(callable(T...): (list<T>)): Closure(): Generator<int, T>
      */
     public function __invoke(): Closure
     {
         return
             /**
              * @param T ...$parameters
              *
-             * @return Closure(callable(mixed|T...): (array<TKey, T>)): Closure(): Generator<int, T>
+             * @return Closure(callable(T...): (list<T>)): Closure(): Generator<int, T>
              */
             static fn (...$parameters): Closure =>
                 /**
-                 * @param callable(mixed|T...): (mixed|array<TKey, T>) $callback
+                 * @param callable(T...): (T|list<T>) $callback
```

Comment:
I don't understand this change.

`$callback` is a user callback, so, it might return anything... no?

Reply:
well, yes and no.

It's a bit tricky to type this one properly to be honest, so it's up to us if we want to put any restrictions on it or not.

From what I see, this is the behaviour:
- the callback can return _any value_, let's call it "x"
- in the next call, "x" is passed to the callback as parameter, with the following behaviour:
         - if "x" _is a list_, it's unpacked and each item in the list will be a different parameter for the callback
         - if "x" _is not a list_, then it means it will be the parameter for the callback, and the callback only takes one argument

Examples of each:

```PHP
$fibonacci = static fn ($a = 0, $b = 1): array =>[$b, $a + $b];
Collection::unfold($fibonacci);

$even = static fn (int $carry): int => $carry + 2;
Collection::unfold($even, -2);
```

The change I made here means that we are linking the return type of the callback with the parameters that are passed in, but it does also mean that the parameters should be the same type. So we're saying that you can either return a list of parameters of the same type OR a single value which should be the same type as the callback param type.

I think this is fine for most use cases. However, if we want to allow more flexibility the only other option is to go full mixed:

```PHP
@param callable(mixed ...): mixed $callback
```

Reply:
I agree. The return type of the closure must be a list in order to fulfill the operation correction.

Maybe we should then get rid of the `(array)` cast and `array_values` then ? WDYT ?

Reply:
yeah those are not needed I think, I'll test it

Reply:
I removed both now and introduced logic to check whether the `$parameters` is a list or a single value. Works the same way as before but makes the code more explicit than the `(array)` cast + `array_values`

Reply:
> Yes it's nice!
> 
> How about being more strict and remove the `is_array()` ?

Oh, are you suggesting to modify the operation to only allow us to return an `array` from the callback?

Reply:
Yes, definitely, that's what this callback is supposed to return, a list.

Reply:
Ok, I'll make the needed changes then. It will be a minor backwards compatibility break

Reply:
I think it might be better to do it now.

Reply:
yeah I'll do it as part of this PR

Reply:
I've updated the code and unit tests, have a look how the usage changes. If you're happy with this I can then go ahead and update the documentation and add some static analysis checks

Reply:
yes definitely :)